code/modules/keybindings/bindings_client.dm

```diff
@@ -1,9 +1,151 @@
+	"R" = 1,
+	"S" = 1,
+	"T" = 1,
+	"U" = 1,
+	"V" = 1,
+	"W" = 1,
+	"X" = 1,
+	"Y" = 1,
+	"Z" = 1,
+	"0" = 1,
+	"1" = 1,
+	"2" = 1,
+	"3" = 1,
+	"4" = 1,
+	"5" = 1,
+	"6" = 1,
+	"7" = 1,
+	"8" = 1,
+	"9" = 1,
+	"-" = 1,
+	"=" = 1,
+	"\[" = 1,
+	"\]" = 1,
+	"\\" = 1,
+	"." = 1,
+	"/" = 1,
+	"`" = 1,
+	"Capslock" = 1,
+	"Numpad0" = 1,
+	"Numpad1" = 1,
+	"Numpad2" = 1,
+	"Numpad3" = 1,
+	"Numpad4" = 1,
+	"Numpad5" = 1,
+	"Numpad6" = 1,
+	"Numpad7" = 1,
+	"Numpad8" = 1,
+	"Numpad9" = 1,
+	"North" = 1,
+	"South" = 1,
+	"East" = 1,
+	"West" = 1,
+	"Northwest" = 1,
+	"Southwest" = 1,
+	"Northeast" = 1,
+	"Southeast" = 1,
+	"Center" = 1,
+	"Return" = 1,
+	"Escape" = 1,
+	"Tab" = 1,
+	"Space" = 1,
+	"Back" = 1,
+	"Insert" = 1,
+	"Delete" = 1,
+	"Pause" = 1,
+	"Snapshot" = 1,
+	"LWin" = 1,
+	"RWin" = 1,
+	"Apps" = 1,
+	"Multiply" = 1,
+	"Add" = 1,
+	"Subtract" = 1,
+	"Divide" = 1,
+	"Separator" = 1,
+	"Decimal" = 1,
+	"Shift" = 1,
+	"Ctrl" = 1,
+	"Numlock" = 1,
+	"Scroll" = 1,
+	"Alt" = 1
+))
+
 // Clients aren't datums so we have to define these procs indpendently.
 // These verbs are called for all key press and release events
 /client/verb/keyDown(_key as text)
 	set instant = TRUE
 	set hidden = TRUE
 
+	client_keysend_amount += 1
+		
+	var/cache = client_keysend_amount
+
+	if(keysend_tripped && next_keysend_trip_reset <= world.time)
+		keysend_tripped = FALSE
+
+	if(next_keysend_reset <= world.time)
+		client_keysend_amount = 0
+		next_keysend_reset = world.time + (1 SECONDS)
+
+	//The "tripped" system is to confirm that flooding is still happening after one spike
+	//not entirely sure how byond commands interact in relation to lag
+	//don't want to kick people if a lag spike results in a huge flood of commands being sent
+	if(cache >= MAX_KEYPRESS_AUTOKICK)
+		if(!keysend_tripped)
+			keysend_tripped = TRUE		
+			next_keysend_trip_reset = world.time + (2 SECONDS)
+		else
+			log_admin("Client [ckey] was just autokicked for flooding keysends; likely abuse but potentially lagspike.")
+			message_admins("Client [ckey] was just autokicked for flooding keysends; likely abuse but potentially lagspike.")
+			QDEL_IN(src, 1)
+			return
+
+	//check if the keycommand is even valid before firing it through the full chain
+	if(!GLOB.valid_keys[_key])
+		//Sanity check, nothing valid in game generates keypress "keys" this long
+		//Means it's some kind of bullshit going on, so get rid of them.
+		if(length(_key) > MAX_KEYPRESS_COMMANDLENGTH)
+			log_admin("Client [ckey] just attempted to send an invalid keypress. Keymessage was over [MAX_KEYPRESS_COMMANDLENGTH] characters, autokicking due to likely abuse.")
+			message_admins("Client [ckey] just attempted to send an invalid keypress. Keymessage was over [MAX_KEYPRESS_COMMANDLENGTH] characters, autokicking due to likely abuse.")
+			QDEL_IN(src, 1)
+			return
+		//No matter what, reject the message.
+		log_admin("Client [ckey] just attempted to send an invalid keypress.  Keymessage was \"[_key]\", report this!.")
```

Comment:
until we know how likely this is I'd suggest just autokicking to preveent spam

I can't hink of very many cases where it will inconvenience people beyond having to rejoin